File: 2024/02/main.R
#!/usr/bin/env -S Rscript --vanilla

read_input <- function(input_file) {
  input_file |>
    readr::read_lines(progress = FALSE)
}

part1 <- function(input) {
  is_safe <- function(input_str) {
    input_ints <- input_str |>
      stringr::str_split_1(" ") |>
      as.integer()
    direction <- ifelse(input_ints[length(input_ints)] > input_ints[1], 1, -1)
    for (i in 2:length(input_ints)) {
      diff <- (input_ints[i] - input_ints[i - 1]) * direction
      if (diff < 1 || diff > 3) {
        return(FALSE)
      }
    }
    TRUE
  }

  input |>
    purrr::map_lgl(is_safe) |>
    table() |>
    purrr::pluck("TRUE")
}


part2 <- function(input) {
  is_safe <- function(input_str, last_chance = FALSE) {
    input_ints <- input_str |>
      stringr::str_split_1(" ") |>
      as.integer()
    direction <- ifelse(input_ints[length(input_ints)] > input_ints[1], 1, -1)
    for (i in 2:length(input_ints)) {
      diff <- (input_ints[i] - input_ints[i - 1]) * direction
      if (diff < 1 || diff > 3) {
        if (last_chance) {
          return(FALSE)
        }
        return(
          is_safe(
            paste(input_ints[-(i - 1)], collapse = " "),
            last_chance = TRUE
          ) || is_safe(
            paste(input_ints[-i], collapse = " "),
            last_chance = TRUE
          )
        )
      }
    }
    TRUE
  }

  input |>
    purrr::map_lgl(is_safe) |>
    table() |>
    purrr::pluck("TRUE")
}

argv <- argparser::arg_parser(
  name = "main.R",
  description = paste(
    "Advent of Code 2024, Day 2:",
    "Take input file containing raw text input, and",
    "print out solution to either part 1 or part 2."
  ),
  hide.opts = TRUE
) |>
  argparser::add_argument(
    arg = "input", help = "Input text file path."
  ) |>
  argparser::add_argument(
    arg = "part", help = "Which part to solve, either 1 or 2.",
    type = "integer"
  ) |>
  argparser::parse_args()

input <- argv$input |>
  read_input()

argv$part |>
  switch(input |> part1() |> cat(sep = "\n"),
    input |> part2() |> cat(sep = "\n")
  )

File: 2024/02/main.py
#!/usr/bin/env python3

import argparse


def part1(input: list[str]) -> int:
    count = 0
    for line in input:
        levels = [int(level) for level in line.split(" ")]
        direction = 1 if levels[-1] - levels[0] > 0 else -1
        safe = True
        for i in range(1, len(levels)):
            diff = (levels[i] - levels[i - 1]) * direction
            if not (1 <= diff <= 3):
                safe = False
                break
        if safe:
            count += 1
    return count


def part2(input: list[str]) -> int:
    def double_check(levels: list[int]) -> bool:
        direction = 1 if levels[-1] - levels[0] > 0 else -1
        for i in range(1, len(levels)):
            diff = (levels[i] - levels[i - 1]) * direction
            if not (1 <= diff <= 3):
                return False
        return True

    count = 0
    for line in input:
        levels = [int(level) for level in line.split(" ")]
        direction = 1 if levels[-1] - levels[0] > 0 else -1
        safe = True
        for i in range(1, len(levels)):
            diff = (levels[i] - levels[i - 1]) * direction
            if not (1 <= diff <= 3):
                safe = double_check(levels[: i - 1] + levels[i:]) or double_check(
                    levels[:i] + levels[i + 1 :]
                )
                if not safe:
                    break
        if safe:
            count += 1
    return count


def main(input_file, part):

    with open(input_file, "r") as f:
        input_str = [line.strip() for line in f.readlines()]

    match part:
        case 1:
            print(part1(input_str))
        case 2:
            print(part2(input_str))


if __name__ == "__main__":
    parser = argparse.ArgumentParser(
        description="""Advent of Code 2024, Day 2:
    Take input file containing raw text input, and
    print out solution to either part 1 or part 2."""
    )
    parser.add_argument("input", type=str, help="Input text file path.")
    parser.add_argument("part", type=int, help="Which part to solve, either 1 or 2.")
    args = parser.parse_args()

    if args.part > 2 or args.part < 1:
        parser.error("`part` argument must be 1 or 2.")

    main(args.input, args.part)
